### PR TITLE
DHFPROD-6284:Throw 4xx error when using fn:error if appropriate

### PR DESCRIPTION
**Pull request by srinathgit**
### Description
Throw 4xx error when using fn:error if appropriate
#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

**Review by ryanjdew — APPROVED** [review]

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/impl/flow-lib.xqy [hunk 1]
```diff
@@ -1023,7 +1026,7 @@ declare function flow:safe-run($func)
     catch($ex) {
       debug:log(xdmp:describe($ex, (), ())),
       trace:error-trace($rfc:item-context, $ex, xdmp:elapsed-time() - $before),
-      fn:error((), "DATAHUB-PLUGIN-ERROR", $ex)
+      httputils:throw-bad-request((), $ex)
```
Comment (srinathgit) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/impl/flow-lib.xqy:1029]: @rjrudin ,
This change and a similar change in flow-lib.sjs in dhfprod-6284 are causing a number of test failures (These failures are for changes in flow-lib.sjs https://jenkins.marklogic.com/blue/organizations/jenkins/Datahub_CI/detail/develop/1692/tests/  alone). Should we not make this change in dhf 4 modules ? I am seeing that no matter a 4.x flow flow succeeds or fails, it always returns a http 200.

Comment (rjrudin) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/impl/flow-lib.xqy:1029]: @srinathgit If a 200 is always returned, then yeah - no need to change these. Can you add a comment at the top of flow-lib.xqy and flow-lib.sjs describing why we're not using the httpUtils functions so future developers are aware?

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy [hunk 2]
```diff
@@ -84,7 +86,7 @@ let $uber-model :=
     (: build uber model with original info :)
     return hent:uber-model((xdmp:to-json($entity-def-map)/object-node(), $supporting-entity-defs)) => map:with("info", $info-map)
   } catch ($e) {
-    fn:error(xs:QName("INVALID-MODEL"), "Unable to validate entity model at URI: " || $trgr:uri || "; cause: " || xdmp:quote($e))
+    httputils:throw-bad-request(xs:QName("INVALID-MODEL"), "Unable to validate entity model at URI: " || $trgr:uri || "; cause: " || xdmp:quote($e))
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy:89]: Can you test this to see what error you get in the client that calls an endpoint that invokes this trigger? I am thinking that the first argument adds no value and should be returned. We don't have some list of error codes that a user can refer to, and returning something that looks like an error code implies that we do keep a list like that. I'm thinking it's less confusing if we just return the error message starting with "Unable to validate".

File: marklogic-data-hub/src/main/resources/ml-modules/services/mlFlow.xqy [hunk 3]
```diff
@@ -69,7 +72,7 @@ declare function get(
           return
             if (fn:exists($flow)) then $flow
             else
-              fn:error((),"RESTAPI-SRVEXERR", (404, "Not Found", "The requested flow was not found"))
+              httputils:throw-not-found((), "The requested flow was not found")
```
Comment (rjrudin) [inline marklogic-data-hub/src/main/resources/ml-modules/services/mlFlow.xqy:75]: Why include "()" as the first arg here?

**Review by rjrudin — CHANGES_REQUESTED** [review]
Main question is whether there's any value in returning the arbitrary error code that some parts of DHF return (and I don't think we should ever pass in "()").

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy (hunk 2 above)
Comment (srinathgit) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy:89]: The first argument is error name as xs:Qname(). The second one is the code which is always "RESTAPI-SRVEXERR"(for 400, 404)
fn:error((xs:QName("FAILED")), "RESTAPI-SRVEXERR",(400,"greater than 50"))
returns 
```
RESTAPI-SRVEXERR: (FAILED) Extension Error: code: 400 message: greater than 50 document: %3
2020-12-07 09:50:37.707 Info:+  <error:code>RESTAPI-SRVEXERR</error:code>
2020-12-07 09:50:37.707 Info:+  <error:name>FAILED</error:name>
2020-12-07 09:50:37.707 Info:+  <error:xquery-version>1.0-ml</error:xquery-version>
2020-12-07 09:50:37.707 Info:+  <error:message>Extension Error: </error:message>
2020-12-07 09:50:37.707 Info:+  <error:format-string>RESTAPI-SRVEXERR: (FAILED) Extension Error:  code: 400 message: greater than 50 document: %3</error:format-string>
....
```
fn:error( (), "RESTAPI-SRVEXERR",(400,"greater than 50")) returns
```
 RESTAPI-SRVEXERR: (err:FOER0000) Extension Error: code: 400 message: greater than 50 document: %3

2020-12-07 09:49:40.410 Info:+  <error:code>RESTAPI-SRVEXERR</error:code>
2020-12-07 09:49:40.410 Info:+  <error:name>err:FOER0000</error:name>
2020-12-07 09:49:40.410 Info:+  <error:xquery-version>1.0-ml</error:xquery-version>
2020-12-07 09:49:40.410 Info:+  <error:message>Extension Error: </error:message>
2020-12-07 09:49:40.410 Info:+  <error:format-string>RESTAPI-SRVEXERR: (err:FOER0000) Extension Error:  code: 400 message: greater than 50 document: %3</error:format-string>
...
```

File: marklogic-data-hub/src/main/resources/ml-modules/services/mlFlow.xqy (hunk 3 above)
Comment (srinathgit) [reply marklogic-data-hub/src/main/resources/ml-modules/services/mlFlow.xqy:75]: In many places we pass error name (first argument for e.g. in SM project), so included that param.

File: marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy (hunk 2 above)
Comment (rjrudin) [reply marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/4/triggers/entity-model-validate-trigger.xqy:89]: Oh I see, I missed that part. Looks good then.

**Review by rjrudin — DISMISSED** [review]

**Comment by srinathgit** [comment]
I have retained the change in /data-hub/4/triggers/entity-model-validate-trigger.xqy as it is being used in 5.x when we deploy entity models. The rest of the changes in 4.x code has been reverted.

**Review by rjrudin — DISMISSED** [review]

**Comment by srinathgit** [comment]
Fixed a failing test

**Review by rjrudin — APPROVED** [review]

**Review by ryanjdew — APPROVED** [review]

**Review by akshaysonvane — APPROVED** [review]

**Pull request merged**